rs/messaging/src/routing/stream_handler.rs

```diff
@@ -735,36 +731,20 @@ impl StreamHandlerImpl {
                 state,
                 available_guaranteed_response_memory,
             ) {
-                Ok(()) => {
+                None => {
                     // Message successfully inducted or dropped.
                     stream.push_accept_signal();
                 }
-                Err((StateError::CanisterMigrating { .. }, RequestOrResponse::Response(_))) => {
-                    // Unable to deliver a response due to migrating canister, push reject signal.
-                    stream.push_reject_signal(RejectReason::CanisterMigrating);
-                }
-                Err((err, RequestOrResponse::Request(request))) => {
-                    // TODO(MR-249): Replace generating reject responses with pushing reject signals.
-                    // Unable to induct a request, generate reject response and push it into `stream`.
-                    let code = match err {
-                        StateError::CanisterNotFound(_) => RejectCode::DestinationInvalid,
-                        StateError::CanisterStopped(_) => RejectCode::CanisterError,
-                        StateError::CanisterStopping(_) => RejectCode::CanisterError,
-                        StateError::CanisterMigrating { .. } => RejectCode::SysTransient,
-                        StateError::QueueFull { .. } => RejectCode::SysTransient,
-                        StateError::OutOfMemory { .. } => RejectCode::CanisterError,
-                        StateError::NonMatchingResponse { .. }
-                        | StateError::BitcoinNonMatchingResponse { .. } => {
-                            unreachable!("Not a user error: {}", err);
-                        }
-                    };
-                    *available_guaranteed_response_memory -=
-                        stream.push(generate_reject_response(&request, code, err.to_string()))
-                            as i64;
-                    stream.push_accept_signal()
-                }
-                Err((_, RequestOrResponse::Response(_))) => {
-                    unreachable!("No signals are generated for response induction failures except for CanisterMigrating");
+                Some((reason, msg)) => {
+                    // Unable to induct a message, push reject signal.
+                    stream.push_reject_signal(reason);
+                    debug_assert!(matches!(
+                        (reason, msg),
+                        // All reasons are legal for requests.
+                        (_, RequestOrResponse::Request(_))
+                            // Only `CanisterMigrating` is legal for responses.
+                            | (RejectReason::CanisterMigrating, RequestOrResponse::Response(_))
+                    ));
```

Comment:
I put this into a `debug_assert!()` because this couldn't fail because of any incoming stream slices or anything like that, it's just the way `induct_message_impl()` is written. We should keep it, put having this fire during tests is good enough.

Reply:
It may be safe, but also pointless. The only way this can fail is if you change the structure in the code and this is caught in tests.